qcodes/instrument_drivers/QDevil/QDevil_QDac_channels.py

```diff
@@ -0,0 +1,833 @@
+        self._get_status_performed = False
+        return snap
+
+    #########################
+    # Channel gets/sets
+    #########################
+
+    def _set_voltage(self, chan, v_set):
+        """
+        set_cmd for the chXX_v parameter
+
+        Args:
+            chan (int): The 1-indexed channel number
+            v_set (float): The target voltage
+
+        If a finite slope has been assigned, we assign a function generator to
+        ramp the voltage.
+        """
+        # validation
+        atten = self.channels[chan-1].vrange.get_latest()
+
+        outofrange = False
+        if v_set > self.vranges[chan][atten]['Max']:
+            v_set = self.vranges[chan][atten]['Max']
+            outofrange = True
+        elif v_set < self.vranges[chan][atten]['Min']:
+            v_set = self.vranges[chan][atten]['Min']
+            outofrange = True
+
+        if outofrange:
+            log.warning('Requested voltage outside reachable range.' +
+                        ' Setting voltage on channel ' +
+                        '{} to {} V. '.format(chan, v_set) +
+                        'Note: _get_latest not up to date')
+
+        slopechans = [sl[0] for sl in self._slopes]
+        if chan in slopechans:
+            slope = [sl[1] for sl in self._slopes if sl[0] == chan][0]
+            # find and assign fg
+            fg = min(self._fgs.difference(set(self._assigned_fgs.values())))
+            self._assigned_fgs[chan] = fg
+            # We need .get and not get_latest in case a ramp was interrupted
+            v_start = self.channels[chan-1].v.get()
+            time = abs(v_set-v_start)/slope
+            log.info('Slope: {}, time: {}'.format(slope, time))
+            # Attenuation compensation and syncing
+            # happen inside _rampvoltage
+            self._rampvoltage(chan, fg, v_start, v_set, time)
+        else:
+            self.write('wav {} 0 0 0;set {} {:.6f}'.format(chan, chan, v_set))
+
+    def _set_vrange(self, chan, switchint):
+        """
+        set_cmd for changing the voltage range: chXX_vrange
+
+        The switchint is an integer. 0: high range, 1: low range.
+
+        The output voltage should be set to zero before changin the range,
+        eventhough we try to keep the voltage output steady if within range.
+        A spike will occur if the voltage is not zero.
+        """
+
+        old = self.channels[chan-1].vrange.get_latest()
+        old_irange = self.channels[chan-1].irange.get_latest()
+
+        if xor(old, switchint):
+            # In the QDevil QDAC version 1.0 we can not have the high current range and the low voltage range at the same time
+            if (old_irange == 1) and (switchint == 1):
+                self.channels[chan-1].irange.set(0)
+                log.warning('QDAC can not be in low voltage range and high current range at the same time. Setting current range to low.')
```

Comment:
I think this is a dangerous concept because it introduces a coupling of parameters (to our experience nobody reads log messages/warnings even errors. Do I understand it correctly that one could introduce a single range parameter, that would both influence the current and voltage?

Reply:
Yes, which is the major reason why I have chosen to refactor and get rid of vrange and irange and introduced "mode". So I was happy with your comment. The new mode parameter has reduced complexity a lot.